File: top_interview_questions/2_medium/2_linked_list/intersection_of_two_linked_lists/solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        while headB:
            headB.is_b = True
            headB = headB.next

        while headA:
            if hasattr(headA, "is_b"):
                return headA

            headA = headA.next

File: top_interview_questions/2_medium/2_linked_list/intersection_of_two_linked_lists/solution.go
package main

func main() {

}

type ListNode struct {
	Val  int
	Next *ListNode
}

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	visited := make(map[*ListNode]bool)

	for headA != nil {
		visited[headA] = true
		headA = headA.Next
	}

	for headB != nil {
		if _, ok := visited[headB]; ok {
			return headB
		}
		headB = headB.Next
	}

	return nil
}
